File: packages/sandbox-runtime/containment_guard.py
"""containment_guard.py — Python port of the action-path ContainmentGuard (PA-ORCH-9).

A 1:1 port of src/lib/orchestrator/containment-guard.ts. A sub-agent declares its tool scopes
up front; before it stages any connector write, the (connector, action) pair is checked here.
Anything not declared is blocked with a typed error — fail closed, symmetric with the web tier.

Scope grammar (identical to the TS guard):
    "*"           -> every connector + action
    "gmail"       -> every action on gmail (== "gmail:*")
    "gmail:*"     -> every action on gmail
    "gmail:send"  -> only the send action on gmail
"""

from __future__ import annotations

from collections.abc import Iterable


class ConnectorScopeError(Exception):
    """Raised when a connector action is outside a sub-agent's declared scopes."""

    def __init__(self, connector: str, action: str) -> None:
        self.connector = connector
        self.action = action
        self.user_message = (
            f"I tried to {action} via {connector}, but that's outside this run's approved "
            f"scope. Want me to ask for the {connector} scope so I can do that?"
        )
        super().__init__(f"ConnectorScope: {connector}:{action} not in declared scopes")


def _norm(value: str) -> str:
    return value.strip().lower()


def scope_token(connector: str, action: str) -> str:
    """Canonical scope token for a connector action."""
    return f"{_norm(connector)}:{_norm(action)}"


def is_action_allowed(connector: str, action: str, declared_scopes: Iterable[str]) -> bool:
    """True iff connector.action is permitted by the declared scopes. Fail closed."""
    c = _norm(connector)
    a = _norm(action)
    if not c or not a:
        return False
    for raw in declared_scopes:
        scope = _norm(raw)
        if not scope:
            continue
        if scope == "*":
            return True
        if scope == c:
            return True
        if scope == f"{c}:*":
            return True
        if scope == f"{c}:{a}":
            return True
    return False


def assert_action_allowed(connector: str, action: str, declared_scopes: Iterable[str]) -> None:
    """Raise ConnectorScopeError when connector.action is out of scope; else return None."""
    scopes = list(declared_scopes)
    if not is_action_allowed(connector, action, scopes):
        raise ConnectorScopeError(connector, action)

File: src/lib/orchestrator/containment-guard.ts
// containment-guard.ts — ContainmentGuard extended to ACTION paths (PA-ORCH-9).
//
// lib/brain/containment-guard.ts guards READS: which brain files a sub-agent may pull into an
// LLM context. This module is the single source of truth for the symmetric WRITE guard: which
// connector actions a sub-agent may attempt. A run declares its `toolScopes` up-front (in
// pa_sub_agent_runs.tool_scopes + the spec). Before any connector write is staged, the
// (connector, action) pair is checked here. Anything not declared is blocked with a typed
// error that surfaces in the chat as "I tried to do X but it's outside my approved scope"
// (§7 criterion 9) — fail closed, symmetric with the read guard.
//
// tool-use.ts (the approval-gate middleware) imports the scope check from here so there is
// exactly ONE ConnectorScopeError class + one matching rule across the orchestrator.
//
// UI label stays "Privacy zones"; we never surface "ContainmentGuard" to customers.

// A scope grants action rights at one of these granularities:
//   "*"            → every connector + action (owner-blessed full-trust runs only)
//   "gmail"        → every action on the gmail connector (== "gmail:*")
//   "gmail:*"      → every action on gmail
//   "gmail:send"   → only the `send` action on gmail
// Scopes are declared on the plan; the dispatcher never widens them at runtime.

export class ConnectorScopeError extends Error {
  readonly connector: string;
  readonly action: string;
  readonly userMessage: string;

  constructor(connector: string, action: string) {
    const userMessage =
      `I tried to ${action} via ${connector}, but that's outside this run's approved scope. ` +
      `Want me to ask for the ${connector} scope so I can do that?`;
    super(`ConnectorScope: ${connector}:${action} not in declared scopes`);
    this.name = "ConnectorScopeError";
    this.connector = connector;
    this.action = action;
    this.userMessage = userMessage;
  }
}

function norm(s: string): string {
  return s.trim().toLowerCase();
}

/** Canonical scope token for a connector action. */
export function scopeToken(connector: string, action: string): string {
  return `${norm(connector)}:${norm(action)}`;
}

/**
 * True iff `connector.action` is permitted by the declared scope set. A scope matches when it
 * is the global wildcard `*`, the bare connector name or `connector:*` (grants all that
 * connector's actions), or the exact `connector:action` pair. Empty/whitespace scopes never
 * match. Case-insensitive. Fail closed.
 */
export function isActionAllowed(
  connector: string,
  action: string,
  declaredScopes: readonly string[],
): boolean {
  const c = norm(connector);
  const a = norm(action);
  if (!c || !a) return false;
  for (const raw of declaredScopes) {
    const scope = norm(raw);
    if (!scope) continue;
    if (scope === "*") return true;
    if (scope === c) return true;
    if (scope === `${c}:*`) return true;
    if (scope === `${c}:${a}`) return true;
  }
  return false;
}

/**
 * Throws ConnectorScopeError when `connector.action` is outside the declared scopes. Returns
 * void when permitted. Call this BEFORE staging any connector write — the gate is structural,
 * upstream of the approval gate, so an out-of-scope action never even reaches the Inbox.
 */
export function assertActionAllowed(
  connector: string,
  action: string,
  declaredScopes: readonly string[],
): void {
  if (!isActionAllowed(connector, action, declaredScopes)) {
    throw new ConnectorScopeError(connector, action);
  }
}

/**
 * Filters a batch of (connector, action) pairs into allowed + blocked. Never throws — for
 * planning-time validation where one out-of-scope action should be surfaced, not abort the
 * whole plan.
 */
export function partitionAllowedActions(
  pairs: readonly { connector: string; action: string }[],
  declaredScopes: readonly string[],
): {
  allowed: { connector: string; action: string }[];
  blocked: { connector: string; action: string }[];
} {
  const allowed: { connector: string; action: string }[] = [];
  const blocked: { connector: string; action: string }[] = [];
  for (const p of pairs) {
    if (isActionAllowed(p.connector, p.action, declaredScopes)) allowed.push(p);
    else blocked.push(p);
  }
  return { allowed, blocked };
}
